File: src/test/java/com/digitalpetri/strictmachine/dsl/ViaBuilderTest.java
package com.digitalpetri.strictmachine.dsl;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import org.junit.jupiter.api.Test;

public class ViaBuilderTest {

  @Test
  void actionBuiltWithViaInstance() throws InterruptedException {
    var fb = new FsmBuilder<State, Event>();

    fb.when(State.S1)
        .on(Event.E1.INSTANCE)
        .transitionTo(State.S2);

    var executed = new AtomicBoolean(false);

    fb.onTransitionFrom(State.S1)
        .to(State.S2)
        .via(Event.E1.INSTANCE)
        .execute(ctx -> executed.set(true));

    assertEquals(State.S2, fb.build(State.S1).fireEventBlocking(Event.E1.INSTANCE));

    assertTrue(executed.get());
  }

  @Test
  void actionBuiltWithViaClass() throws InterruptedException {
    var fb = new FsmBuilder<State, Event>();

    fb.when(State.S1)
        .on(Event.E1.INSTANCE)
        .transitionTo(State.S2);

    var executed = new AtomicBoolean(false);

    fb.onTransitionFrom(State.S1)
        .to(State.S2)
        .via(Event.E1.class)
        .execute(ctx -> executed.set(true));

    assertEquals(State.S2, fb.build(State.S1).fireEventBlocking(Event.E1.INSTANCE));

    assertTrue(executed.get());
  }

  @Test
  void actionBuiltWithViaPredicate() throws InterruptedException {
    var fb = new FsmBuilder<State, Event>();

    fb.when(State.S1)
        .on(Event.E1.INSTANCE)
        .transitionTo(State.S2);

    var executed = new AtomicBoolean(false);

    fb.onTransitionFrom(State.S1)
        .to(State.S2)
        .via(e -> e instanceof Event.E1)
        .execute(ctx -> executed.set(true));

    assertEquals(State.S2, fb.build(State.S1).fireEventBlocking(Event.E1.INSTANCE));

    assertTrue(executed.get());
  }

  @Test
  void actionBuiltWithViaAny() throws InterruptedException {
    var fb = new FsmBuilder<State, Event>();

    fb.when(State.S1)
        .onAny()
        .transitionTo(State.S2);

    var executed = new AtomicBoolean(false);

    fb.onTransitionFrom(State.S1)
        .to(State.S2)
        .viaAny()
        .execute(ctx -> executed.set(true));

    for (Event event : List.of(new Event.E1(), new Event.E2(), new Event.E3())) {
      executed.set(false);
      assertEquals(State.S2, fb.build(State.S1).fireEventBlocking(event));
      assertTrue(executed.get());
    }
  }
  
}

File: src/test/java/com/digitalpetri/strictmachine/dsl/ViaBuilderTest.kt
/*
 * Copyright 2018 Kevin Herron
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.digitalpetri.strictmachine.dsl

import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.util.concurrent.atomic.AtomicBoolean

class ViaBuilderTest {

    @Test
    fun `Action built with via Instance`() {
        val fb = FsmBuilder<State, Event>()

        fb.`when`(State.S1)
            .on(Event.E1)
            .transitionTo(State.S2)

        val executed = AtomicBoolean(false)

        fb.onTransitionFrom(State.S1)
            .to(State.S2)
            .via(Event.E1)
            .execute { executed.set(true) }

        fb.build(State.S1).fireEventBlocking(Event.E1)

        assertTrue(executed.get()) {
            "expected execute callback to set executed=true"
        }
    }

    @Test
    fun `Action built with via Class`() {
        val fb = FsmBuilder<State, Event>()

        fb.`when`(State.S1)
            .on(Event.E1)
            .transitionTo(State.S2)

        val executed = AtomicBoolean(false)

        fb.onTransitionFrom(State.S1)
            .to(State.S2)
            .via(Event.E1::class.java)
            .execute { executed.set(true) }

        fb.build(State.S1).fireEventBlocking(Event.E1)

        assertTrue(executed.get()) {
            "expected execute callback to set executed=true"
        }
    }

    @Test
    fun `Action built with via Predicate`() {
        val fb = FsmBuilder<State, Event>()

        fb.`when`(State.S1)
            .on(Event.E1)
            .transitionTo(State.S2)

        val executed = AtomicBoolean(false)

        fb.onTransitionFrom(State.S1)
            .to(State.S2)
            .via { e -> e == Event.E1 }
            .execute { executed.set(true) }

        fb.build(State.S1).fireEventBlocking(Event.E1)

        assertTrue(executed.get()) {
            "expected execute callback to set executed=true"
        }
    }

    @Test
    fun `Action built with via Any`() {
        val fb = FsmBuilder<State, Event>()

        fb.`when`(State.S1)
            .onAny()
            .transitionTo(State.S2)

        val executed = AtomicBoolean(false)

        fb.onTransitionFrom(State.S1)
            .to(State.S2)
            .viaAny()
            .execute { executed.set(true) }

        for (event in listOf(
            Event.E1,
            Event.E2,
            Event.E3,
            Event.E4
        )) {
            executed.set(false)

            fb.build(State.S1).fireEventBlocking(event)

            assertTrue(executed.get()) {
                "expected execute callback to set executed=true"
            }
        }
    }

}
